prow/clonerefs/options.go

```diff
@@ -0,0 +1,177 @@
+	DefaultGitUserName  = "ci-robot"
+	DefaultGitUserEmail = "ci-robot@k8s.io"
+)
+
+// ResolveOptions will resolve the set of options, preferring
+// to use the full JSON configuration variable but falling
+// back to user-provided flags if the variable is not
+// provided.
+func ResolveOptions() (*Options, error) {
+	options := &Options{}
+	if jsonConfig, provided := os.LookupEnv(JSONConfigEnvVar); provided {
+		if err := json.Unmarshal([]byte(jsonConfig), &options); err != nil {
+			return options, fmt.Errorf("could not resolve config from env: %v", err)
+		}
+		return options, nil
+	}
+
+	flag.StringVar(&options.SrcRoot, "src-root", "", "Where to root source checkouts")
+	flag.StringVar(&options.Log, "log", "", "Where to write logs")
+	flag.StringVar(&options.GitUserName, "git-user-name", DefaultGitUserName, "Username to set in git config")
+	flag.StringVar(&options.GitUserEmail, "git-user-email", DefaultGitUserEmail, "Email to set in git config")
+
+	refs := gitRefs{}
+	aliases := pathAliases{}
+	flag.Var(&refs, "repo", "Mapping of Git URI to refs to check out, can be provided more than once")
+	flag.Var(&aliases, "clone-alias", "Mapping of org and repo to path to clone to, can be provided more than once")
+	flag.Parse()
+
+	options.GitRefs = refs.gitRefs
+
+	for _, pathAlias := range aliases.aliases {
+		for _, ref := range options.GitRefs {
+			if pathAlias.Org == ref.Org && pathAlias.Repo == ref.Repo {
+				ref.PathAlias = pathAlias.Path
+			}
+		}
+	}
+	return options, nil
+}
+
+type gitRefs struct {
+	gitRefs []*kube.Refs
+}
+
+func (r *gitRefs) String() string {
+	representation := bytes.Buffer{}
+	for _, ref := range r.gitRefs {
+		fmt.Fprintf(&representation, "%s,%s=%s", ref.Org, ref.Repo, ref.String())
+	}
+	return representation.String()
+}
+
+// Set parses out a kube.Refs from the user string.
+// The following example shows all possible fields:
+//   org,repo=base-ref:base-sha[,pull-number:pull-sha]...
+// For the base ref and every pull number, the SHAs
+// are optional and any number of them may be set or
+// unset.
+func (r *gitRefs) Set(value string) error {
+	gitRef, err := ParseRefs(value)
+	if err != nil {
+		return err
+	}
+	r.gitRefs = append(r.gitRefs, gitRef)
+	return nil
+}
+
+type pathAliases struct {
+	aliases []ClonePathAlias
+}
+
+func (a *pathAliases) String() string {
+	representation := bytes.Buffer{}
+	for _, resolver := range a.aliases {
+		fmt.Fprint(&representation, resolver.String())
+	}
+	return representation.String()
+}
+
+// Set parses out path aliases from user input
+func (a *pathAliases) Set(value string) error {
+	resolver, err := ParseAliases(value)
+	if err != nil {
+		return err
+	}
+	a.aliases = append(a.aliases, resolver)
+	return nil
+}
+
+// Encode will encode the set of options in the format that
+// is expected for the configuration environment variable
+func Encode(options Options) (string, error) {
+	encoded, err := json.Marshal(options)
+	return string(encoded[:]), err
```

Comment:
What does the `[:]` do here?

Reply:
I got confused and thought we had a byte array, not a slice. Will remove.